### DISCUSSION TITLE
elm-pages 2.0 public beta instructions and feedback!

### DISCUSSION THREAD

**Discussion by dillonkearns** [dillonkearns/elm-pages#188] [General]
I'm officially opening up the elm-pages 2.0 public beta. Here are [the instructions for setting up a fresh elm-pages 2.0 beta project](https://github.com/dillonkearns/elm-pages/blob/static-files/docs/2.0-beta.md).

Let's use this thread to track feedback. Please comment with anything you find to be confusing, missing, or that could be improved. I'd love to hear any and all thoughts here!

## My Current Todo List
- [ ] Update meta tags in docs site
- [x] Remove lorem ipsum text for lower icons section in landing page
- [x] Update code snippets on landing page to reflect recent API changes
- [x] Finish package docs for
	- [x] Glob
	- [x] QueryParams
	- [x] DataSource.File
	- [x] Dedupe/clean up docs for DataSource & DataSource.Http
	- [x] Pages.PageUrl
- [x] Cache showcase images (either DPR or cloudinary)

**Comment by wolfadex** [comment]
When creating a dynamic page the `RouteParams` doesn't place a space between arguments.

See the `String,splat` which should be `String, splat`.
<img width="436" alt="Screen Shot 2021-05-24 at 3 42 26 PM" src="https://user-images.githubusercontent.com/1542461/119410828-a8f2a780-bca6-11eb-8acc-d522d22bd6ee.png">

Super minor but probably a nice little polish thing.

**Reply by dillonkearns** [reply]
Good catch! I agree, those little details make things feel more solid and are definitely worth polishing 👍

**Comment by iffies** [comment]
When adding a page with the following command (assuming we must keep using the `npx` command during the beta) with `npx dillonkearns/elm-pages#8ccdf30 add Sponsors`, its `view` function is generated as:

```elm
view : StaticPayload Data RouteParams -> Document Msg
view static =
    Document.placeholder "Sponsors"
```

Should this be `View` instead of `Document`? If so, `Page.buildNoState` demands a more nuanced definition of `view`, similar to that provided by the autogenerated `Index.elm` file: 

```elm
view :
    Maybe PageUrl
    -> Shared.Model
    -> StaticPayload Data RouteParams
    -> View Msg
view maybeUrl sharedModel static =
    View.placeholder "Sponsors"
```

Curious if `Document` is legacy naming or if you're encouraging developers to create a custom wrapper for `View`.

**Reply by dillonkearns** [reply]
Oh yes, good catch! I changed the name from `Document` to `View`. I pushed a fix for the `elm-pages add` command.

You can run it with the latest commit there: `npx dillonkearns/elm-pages#badd971d add Sponsors`.

And by the way, the reason I use a specific hash instead of a branch name is because unfortunately `npx` will cache the branch at a fixed point. So at least doing it with a commit hash removes the confusion around being stuck on an old version implicitly. I wish there was an easy way to force it to use the latest version 🤷‍♂️

**Comment by ringods** [comment]
Hello @dillonkearns, you already pointed me to some video resources on the 2.0 beta which are most helpful.

One thing I still want to mention as a newbie is this: in the videos and the examples, every page template(?) calls into the `Page` module, e.g. `Page.singleRoute`.

The `Page` module is not in the package, which means there is no documentation generated for it:

https://elm-doc-preview.netlify.app/?repo=dillonkearns%2Felm-pages&version=static-files

The module is in the generated (git-ignored) `.elm-pages` folder inside my project setup. 

Will this change for the final 2.0?

**Reply by dillonkearns** [reply]
Hello @ringods, thank you for the feedback!

This is a challenging one because the `Pages.elm` module is generated and available locally in order to minimize the number of type variables you have to use. If `Page.elm` wasn't a generated file, then it would need type variables to reference the following types:

- `View.View`
- `Shared.Model`
- `Shared.SharedMsg`
- `Shared.Data`

So moving it into the published package would mean it would be more verbose to create type annotations, and in some cases the error messages would be less helpful because those types can't be anything, they must be exactly referring to those type definitions in the user's codebase.

I've just pushed up an experiment to try a local docs page, but it does have some shortcomings. You can try it out if you update to the latest version (should be something like `cd elm-pages && git pull` if you have an existing elm-pages 2.0 beta setup).

If you run `npm install && npx elm-pages docs` (you might need to npm install to get the new dependency I added), that command gives a local docs preview using https://github.com/dmy/elm-doc-preview. It looks like this:

![Image 2021-05-28 at 11 37 44 AM](https://user-images.githubusercontent.com/1384166/120028174-36850e80-bfa9-11eb-9f0c-631d94423577.jpg)

The main shortcoming is that if you have a compiler error, you won't see the docs until you resolve that error. So let's consider this an experimental idea given the known limitation. I'm open to other ideas to make it easy to navigate the generated APIs, but unfortunately I think there are no easy answers but more clever solutions with different tradeoffs.

**Reply by dillonkearns** [reply]
Also, it's worth noting that when this is a published Elm package, I believe this sidebar will only display the generated modules, not all of the `elm-pages` module docs. The `elm-pages` docs from the published package would be available by clicking to them under the "dependencies" section. Not a bad thing I think, but worth noting.

**Comment by MarcoDaniels** [comment]
Hi, I found an interesting issue.

When using `DataSource` to create routes dynamically, if the url from the data source starts with a slash `/about-us` it will display a "Page not found." when running the dev server. But it works when running build.

The same happens when just doing something like:
```
routes =
    DataSource.succeed [ { slug = "/about-us" } ]
```

**Reply by dillonkearns** [reply]
That's a good catch, thanks for reporting that!

I'm thinking it might make sense to just URL encode those route params. It's a little confusing either way, but it is valid to have `%2F` (an encoded `/`) in a URL, so there should be a way to express that.

**Comment by tennety** [comment]
<img width="704" alt="File-Based Routing - elm-pages docs 2021-06-09 22-28-02" src="https://user-images.githubusercontent.com/147759/121455230-17f02180-c972-11eb-9e59-7b83b719ec6b.png">

Found what looks like a tiny typo while going through the docs.

Also, question: do you already have docs for an upgrade path from v1 (I'm using the starter repo)? I'm thinking it would be getting rid of `src` and `gen` and bootstrapping a fresh project.

**Reply by tennety** [reply]
My bad, found the docs within the submodule. You think of everything!

**Reply by dillonkearns** [reply]
Fixed the typo, thanks!

**Reply by dillonkearns** [reply]
> Also, question: do you already have docs for an upgrade path from v1 (I'm using the starter repo)? I'm thinking it would be getting rid of src and gen and bootstrapping a fresh project.

I'm working on migration instructions here: https://github.com/dillonkearns/elm-pages/blob/static-files/docs/migrating-from-1-to-2.md. If there are some particular questions that come up, let me know and I'll try to add some details about it in that document!

**Comment by bdukes** [comment]
I'm running this on Windows, and when I call `npx dillonkearns/elm-pages#static-files add Articles.Category_`, the generated code has, e.g., `import Page.Articles\Category_` instead of `import Page.Articles.Category_`. All throughout `Route.elm` and `TemplateModulesBeta.elm` it's being included as `Articles\Category_`.

Looks like maybe you should be splitting on [`path.sep`](https://nodejs.org/dist/latest-v14.x/docs/api/path.html#path_path_sep) instead of `"/"` https://github.com/dillonkearns/elm-pages/blob/aa0381c13ded7f9ca5a6a6f8eed8714fca57bf5a/generator/src/generate-template-module-connector.js#L13

**Reply by dillonkearns** [reply]
Good catch! Thanks for pointing me to that line of code. I just pushed up a change. Could you try out the latest one and let me know how it goes?

Also, I've noticed that the npx command sometimes caches the branch's commit and doesn't fetch the latest one, so you may need to point to the latest commit, 22a3d88f6307cbf7323ed238175b96a46f108038, instead of the branch.

**Reply by bdukes** [reply]
That fixed it, thanks for the quick reply!

**Reply by dillonkearns** [reply]
Fantastic! Thanks so much for reporting the issue and testing it out 🙏
